### ISSUE TITLE
Enable Customer Comparer for IEnumerable<T>.Should().Contain()/NotContain()

### ISSUE THREAD

**Issue by DerAlbertCom** [fluentassertions/fluentassertions#699]
I recently used 
```
    ResultIEnumerable<Claim>.Should().Contain(ExcptedIEnumerable<Claim>);
```
The problem with the `System.Security.Claims.Claim` that sometimes it is only needed to compare the Value and Type property of the Claim. And not not ALL properties (this may also true for other examples).

Currently i write a class in that case, which also implements **Equals()**.
```
    public class ClaimTestModel
    {
        public string Type { get; }
        public string Value { get; }

        public ClaimTestModel(string type, string value)
        {
            Type = type;
            Value = value;
        }

        public override string ToString()
        {
            return $"\n{{ {Type} = {Value} }}";
        }

        protected bool Equals(ClaimTestModel other)
        {
            return string.Equals(Type, other.Type) && string.Equals(Value, other.Value);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((ClaimTestModel) obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return ((Type != null ? Type.GetHashCode() : 0) * 397) ^ (Value != null ? Value.GetHashCode() : 0);
            }
        }
    }
```

Also, i have to write a helper method for converting Claim to ClaimTestModel

```
    public static class ClaimModelExtension
    {
        public static IEnumerable<ClaimTestModel> ToCompareModel(this IEnumerable<Claim> claims)
        {
            foreach (var claim in claims)
            {
                yield return new ClaimTestModel(claim.Type, claim.Value);
            }
        }
    }
```

And this only while i need to compare only the two properties in this particular test.
What i really want is to provide a simple IComparer, or predicate which is used on every items. 
This is easy and small.
There is currently a possibility to add a predicate as expected, but this in not suitable for comparing a complete collection. It is possible to write an extension methods, for an easier syntax for this. But with the extension method it is not easy to get the correct error message for the concrete missing instance. So i think this should be supported natively.

**Comment by MJLHThomassen-Eurocom** [comment]
This is already possible in 5.0.0-beta0.0.4

collection.Should().Contain(x => x.Value == someValue && x.OtherValue == someOtherValue);

**Comment by DerAlbertCom** [comment]
No, this only checks if  one  element is on the list.

I would like to compare to lists. With Instances which are not implementing Equals()

**Comment by dennisdoomen** [comment]
If you're comparing two lists and want to include certain properties, use `collection.Should().BeEquivalentTo`.

**Comment by DerAlbertCom** [comment]
@dennisdoomen  no, this does not work in that case (System.Security.Claims.Claims), because not alle properties are needed in every case, and are useless. Also, the message which is given, does not help. In this test, only Type and Value are relevant.

Xunit.Sdk.XunitException: Expected collection 

`{amr: pwd, given_name: Albert, family_name: Weinert, http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier: 4711-0815, name: albert.weinert@aaa.software, email: albert.weinert@aaa.software}`

to be equivalent to 

`{given_name: Albert, family_name: Weinert, amr: pwd, http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier: 4711-0815, name: albert.weinert@aaa.software, email: albert.weinert@aaa.software}, `

but it misses 

`{given_name: Albert, family_name: Weinert, amr: pwd, http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier: 4711-0815, name: albert.weinert@aaa.software, email: albert.weinert@aaa.software}.`

**Comment by DerAlbertCom** [comment]
also, i expect that BeEquivalentTo() als compares against the count of the items,  but this is different than Contains().

**Comment by dennisdoomen** [comment]
What about `collection.Should().Contain(item => item.Value == myValye && item.Type == myType)`?

**Comment by DerAlbertCom** [comment]
Now we go in circles. ;)

**Comment by DerAlbertCom** [comment]
This is not practical if i would like to see if a couple of elements are in the list

**Comment by dennisdoomen** [comment]
Ah, you want to have a sub-set comparison. Then no, there's no such API.

**Comment by DerAlbertCom** [comment]
@dennisdoomen  that's why i opened this issue ;) To discuss a possible solution.

**Comment by dennisdoomen** [comment]
So you're at the right place. I just don't have a satisfying answer for you ;-)

**Comment by jnyrup** [comment]
To support this, we could e.g. to take an `IEqualityComparer<T>` as parameter to compare elements for equality.

I.e. introduce the following methods in the API:
```csharp
Equal(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)
NotEqual(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)

StartWith(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)
EndWith(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)

Contain(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)
NotContain(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)

BeSubsetOf(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)
NotBeSubsetOf(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)

IntersectWith(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)
NotIntersectWith(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)
```

If my memory serves me well, the collection comparison code is quite coupled to using `Equals(object obj)` for handling non-generic collections.
So it would probably require a larger overhaul of the collection code.

**Comment by dennisdoomen** [comment]
Or a `Should().ContainEquivalentOf()` which then uses the structural equvalency API to do a sub-set comparison.

**Comment by DerAlbertCom** [comment]
Yes and No. I like this for using anonymous type, which is fine for ONE Instance, not for multiple instance. and for multiple instance I (me and myself) prefer a concrete class, and for this i would have to implement a custom type. But this is the main point of this issue ;) But this may only me.

**Comment by dennisdoomen** [comment]
Not sure I understand you. It could be based on `Should().BeEquivalentTo`  which supports all the bells and whistles to decide what you want to include in the comparison and how to compare properties.

**Comment by DerAlbertCom** [comment]
Oh, if i can put a comparer in there ... i'm good with that ;) So i didn't understand you.

**Comment by james-ecargo** [comment]
I'm not sure if I can divine exactly what the solution is here. Does this mean there an enhancement coming that will allow the caller to specify an `IEqualityComparer<T>` instance for methods like `Contains`?

..and if that is the case then does FA contain an easily accessible structural equivalency equality comparer that could be passed in?

**Comment by AroglDarthu** [comment]
I agree that the collection comparison methods could really use overloads that take an IEqualityComparer<T>. Do you have a roadmap for that feature?

**Comment by jnyrup** [comment]
@james-ecargo @AroglDarthu 
Fluent Assertions is a spare time project, heavily driven by user contributions.
Unless an issue/PR is explicitly marked with a [milestone](https://github.com/fluentassertions/fluentassertions/milestones), there are no promises about if/when a suggestion will be available.

My earlier comment about API changes, was meant as suggestions in two ways:
* As a general discussion of how the API *could* look like,
* and *if* consensus, as hints to whoever would pick up the task of implementing it.

**Comment by DerAlbertCom** [comment]
Just to throw in an alternate syntax for the suggestion of @jnyrup 
```
Contain<TComparer>(IEnumerable<T> other,  string because = "", params object[] becauseArgs) where TComparer : IEqualityComparer<T>, new()
```
I don't think i like that and it is contrary to the normal usage of comparers. So, i think we should go with the suggestions.

If no one is agains the syntax i will give them a spin next week.

Starting with 
```
Contain(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)
NotContain(IEnumerable<T> other, IEqualityComparer<T> comparer, string because = "", params object[] becauseArgs)
```

To see if I get the code and error messages right.

**Comment by DerAlbertCom** [comment]
An alternative way would could be
```
ContainEquivalentsTo(IEnumerable other, string because = "", params object[] becauseArgs)
NotContainEquivalentsTo(IEnumerable other, string because = "", params object[] becauseArgs)
```
With that there my be better error messages possible.

**Comment by smadurange** [comment]
I think it would be good to be able to pass a custom equality comparer for any function that does an equality comparison. It's quite limiting when the functions only use the default comparer.

**Comment by dennisdoomen** [comment]
> Yes and No. I like this for using anonymous type, which is fine for ONE Instance, not for multiple instance. and for multiple instance I (me and myself) prefer a concrete class, and for this i would have to implement a custom type. But this is the main point of this issue ;) But this may only me.

Since 5.0, you can compare your collection of claims against a collection of anonymous types that only define the properties you care about.

**Comment by dennisdoomen** [comment]
> I think it would be good to be able to pass a custom equality comparer for any function that does an equality comparison. It's quite limiting when the functions only use the default comparer.

It would require a significant number of overloads here and there.

**Comment by jukkahyv** [comment]
Instead of changing the existing methods, what about adding a new call between Should() and Contain() which specifies a custom equality comparer? It could be repeated multiple times.

```csharp
ResultIEnumerable<Claim>.Should().UsingEqualityComparer(new MyClaimComparer()).Contain(ExcptedIEnumerable<Claim>);
```

**Comment by AroglDarthu** [comment]
I think that would add some obfuscation to the API. My preference would be to add overloads.

**Comment by AroglDarthu** [comment]
>Since 5.0, you can compare your collection of claims against a collection of anonymous types that only define the properties you care about.

Sweet, but if re-implementing the custom equality comparer logic in assertions could be prevented by allowing those equality comparers to be used, then that would still be preferable.

**Comment by dennisdoomen** [comment]
The original question can be solved with :

```
actualClaims.Should().BeEquivalentTo(expectedClaims, opt => opt
  .Including(x => x.Type)
  .Including(x => x.Value));
```

**Comment by tg73** [comment]
Same/similar problem for me: I need to compare instances of Orleans grain interfaces. The comparable identity of such instances can only be accessed through a property of an object returned by an extension method on the grain interface instance:
```
ISomeGrain g = ...;
var identity = g.GetGrainIdentity().IdentityString;
// GetGrainIdentity() is an extension method on IGrain, all grain interfaces must implement IGrain.
```
This way of obtaining identity is part of Microsoft's Orleans framework, I have no control over it. If I try to use `opt.Including( x => x.GetGrainIdentity().IdentityString )` I get `System.ArgumentException : Expression <x.GetGrainIdentity().IdentityString> cannot be used to select a member. (Parameter 'expression')`. All grains implement `IGrain`, and I have a custom equality comparer for `IGrain`. I don't yet see how I can solve this elegantly, without lots of code repetition, in FluentAssertions. At present, I use LINQ to produce a collection of `string` from `GetGrainIdentity().IdentityString`, and then work with that - not very elegant.

**Comment by dennisdoomen** [comment]
You can either use the `When`/`Using` construct as explained [here](https://fluentassertions.com/objectgraphs/#equivalency-comparison-behavior) or add a custom implementation of `IEquivalencyStep` as described [here](https://fluentassertions.com/extensibility/#equivalency-assertion-step-by-step).

**Comment by jukkahyv** [comment]
Actually, regarding my suggestion, it would also work if you could supply an EqualityComparer for EquivalencyAssertionOptions, or is it already possible? I'm just saying because EqualityComparer is the "standard" .NET way of doing it, and often there is already a suitable equality comparer available that could be used as is. I found this post searching by FluentAssertions + EqualityComparer by the way.

**Comment by tg73** [comment]
@dennisdoomen Thanks for the helpful info. I've now read the links suggested, and had previously tried the `Using` construct, but had two points of confusion:

1. I could not work out how to represent the logic I need when anchored to `ctx.Subject.Should()`, which seems to be a requirement of the FA API (as far as I can understand it). How should I express the condition `subjectGrain.GetGrainIdentity().IdentityString == expectedGrain.GetGrainIdentity().IdentityString` so that it can be used as the argument to `Using(...)`?

2. I didn't quite understand how the `Using` construct would work when comparing two sequences (regardless of order) with `BeEquivalentTo`, because it seems to have no way of expressing whether a pair of elements simply match or not (ie, returning a boolean instead of throwing). Maybe there's some magic at work? If so, is there any documentation which covers FA's fundamental principles of operation, which might help me to understand better?

**Comment by dennisdoomen** [comment]
> Actually, regarding my suggestion, it would also work if you could supply an EqualityComparer for EquivalencyAssertionOptions, or is it already possible?

I like that.

**Comment by dennisdoomen** [comment]
> 1. I could not work out how to represent the logic I need when anchored to `ctx.Subject.Should()`, which seems to be a requirement of the FA API (as far as I can understand it). How should I express the condition `subjectGrain.GetGrainIdentity().IdentityString == expectedGrain.GetGrainIdentity().IdentityString` so that it can be used as the argument to `Using(...)`?

What about `Using<IGrain>(ctx => ctx.Subject.GetGrainIdentity().IdentityString.Should().Be(ctx.Expectation.GetGrainIdentity().IdentityString()).WhenTypeIs<IGrain>()`?
